qcodes_contrib_drivers/drivers/SingleQuantum/SingleQuantum.py

```diff
@@ -0,0 +1,498 @@
+            TCP_IP_PORT=12345,
+            CNTS_BUFFER=100):
+        threading.Thread.__init__(self)
+        self.lock = threading.Lock()
+        self.rlock = threading.RLock()
+        self.TCP_IP_ADR = TCP_IP_ADR
+        self.TCP_IP_PORT = TCP_IP_PORT
+
+        self.socket = socket.socket(
+            socket.AF_INET, socket.SOCK_STREAM)
+        self.socket.connect((self.TCP_IP_ADR, self.TCP_IP_PORT))
+        # self.socket.settimeout(.1)
+        self.BUFFER = 1000000
+        self.shutdown = False
+
+        self.cnts = []
+        self.CNTS_BUFFER = CNTS_BUFFER
+        self.n = 0
+
+    @synchronized_method
+    def close(self):
+        # print("Closing Socket")
+        self.socket.close()
+        self.shutdown = True
+
+    @synchronized_method
+    def get_n(self, n):
+        n0 = self.n
+        while self.n < n0 + n:
+            time.sleep(0.001)
+        cnts = self.cnts
+        return cnts[-n:]
+
+    def run(self):
+        while self.shutdown is False:
+            if sys.version_info.major == 3:
+                data_raw = str(self.socket.recv(self.BUFFER), 'utf-8')
+            elif sys.version_info.major == 2:
+                data_raw = self.socket.recv(self.BUFFER)
+
+            data_newline = data_raw.split('\n')
+
+            v = []
+            for d in data_newline[0].split(','):
+                v.append(float(d))
+
+            with self.lock:
+                self.cnts.append(v)
+                # Keep Size of self.cnts
+                len_ = len(self.cnts)
+                if len_ > self.CNTS_BUFFER:
+                    self.cnts = self.cnts[len_ - self.CNTS_BUFFER:]
+                self.n += 1
+
+
+class ChannelArray(ParameterWithSetpoints):
+    """Fetches the correct row from the matrix counters,  based on the channel
+    This is a parameter with setpoints, where the setpoints are the time stamps
+    Args:
+        channel (int): the channel number to select the row from the counters matrix
+    Return (numpy_array): a numpy vector of length npts, containing the counts
+    """
+
+    def __init__(self, channel, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._channel = channel
+
+    def get_raw(self):
+        return self.root_instrument.counters.get_latest()[
+            self._channel]
+
+
+class TimeArray(Parameter):
+    """Fetches the row with time stamps, currently channel 0
+    This parameter is used as the setpoints for the channelarray
+    Args:
+        channel (int): the channel number to select the row from the counters matrix (channel 0 is the timestamp)
+    Return (numpy_array): a numpy vector of length npts, containing the time stamps
+    """
+
+    def __init__(self, channel, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._channel = channel
+
+    def get_raw(self):
+        return np.array(self.root_instrument.counters.get_latest()[
+            self._channel])
+
+
+class WebSQControlqcode(IPInstrument):
```

Comment:
As far as I can see all the communcation goes over your own talk implementation. As such there seems to be no use of the IPInstrument and you might as well inherit direcectly from the Instument class

Reply:
Implemented suggestion, everything should still work.

A question, is there a way to do JSON communication over tcp/ip natively in QCoDeS?

Reply:
No not at the moment. If there are a larger number of such instrumentes it may make sense to create such an instrument class